### ISSUE TITLE
Sub tasks

### ISSUE THREAD

**Issue by animalillo** [owncloud/tasks#49]
Would be nice to have something like little tasks inside the tasks that compose the task.

**Comment by nodiscc** [comment]
IMHO the tasks app should stay simple. @animalillo if you want to have sub-tasks, you can use a keyword for each task in your main project, for example `@work finish stuff` `@work buy paper` `@work report`, `@diy draw plans`. This way it remains searchable/filterable and there's no need to add a feature most users won't use.

@raimund-schluessler what do you think? wontfix?

**Comment by animalillo** [comment]
I dont think that adding sub tasks would be so hard or complicated, a way of adding checkboxes would be good enough, even if it's like github does, that would be perfect for me.

A task that haves several steps to be completed its a really common type of task! i think any task app should have that feature

**Comment by nodiscc** [comment]
@animalillo can you use the task description field for this? Keep in mind that the tasks app should follow the CalDav tasks spec (VTODO) and I can't find a specification for sub-tasks. How do you expect it to look/work?

**Comment by raimund-schluessler** [comment]
I already thought about this a bit. Sub-Tasks could be implemented using a non-standard property like X-OC-SUBTASK (see http://tools.ietf.org/html/rfc5545#section-3.8.8.2). So there could be a list of sub-tasks shown in the main task, although they would only be shown inside the task app, as no other app would know this property (except it would be implemented in particular).

However, I think this is a feature worth considering.

**Comment by animalillo** [comment]
@raimund-schluessler yes, that would be perfect, and I would really appreciate it, its really handy to have sub-tasks.

One app I liked its trello, but for little tasks or non collaborative ones it's just too much, plus it's outside owncloud hehe

**Comment by DemoFreak** [comment]
> ... I would really appreciate it, its really handy to have sub-tasks.

+1

> I already thought about this a bit. Sub-Tasks could be implemented using a non-standard property like X-OC-SUBTASK

Other Apps (i.e. KDE Kontact, Android Mirakel, maybe much others) also have the ability to manage sub-tasks. Is there any way to keep the sub-task property across sync borders or is this completely proprietary?

**Comment by raimund-schluessler** [comment]
Well, in the icalendar specification https://www.ietf.org/rfc/rfc2445.txt I didn't find anything regarding sub-tasks. But if other apps with CalDAV support feature sub-tasks we could just have a look at the calendar entries they are generating and which properties they are using for sub-tasks.

**Comment by DemoFreak** [comment]
How can I help there? Export Akonadi/Kontact calendar to .ics file and grep for VTODO entries?
Sorry, I'm a simple user and no developer. ;-)

**Comment by Bullnados** [comment]
Sub Task would be very nice. +1
Perhaps like in todoist.com and not in wunderlist3 <3

**Comment by raimund-schluessler** [comment]
I don't know todoist and don't really want to create an account there. Could you post a screenshot or describe what you mean by "like in todoist.com and not in wunderlist3"?

**Comment by raimund-schluessler** [comment]
@DemoFreak Thanks, very interesting. This could be the way to go.

**Comment by Bullnados** [comment]
Here is an example. With the three dashes u can move a normal task to a sub- or subsubtask. Same u can do with Lists/Projects.
On the left side the Lists/projects u can select the Mainproject, than u have first screenshot.
![example_todoist](https://cloud.githubusercontent.com/assets/11041576/6236334/327a7614-b6ea-11e4-82a6-197a0600d100.png)
Or just select subproject than u have second screenshot
![example2_todoist](https://cloud.githubusercontent.com/assets/11041576/6236346/4c19ca20-b6ea-11e4-9395-341be8e325d0.png)

**Comment by xygrec** [comment]
A very nice implementation is in Kolab's roundcube plugin: http://git.kolab.org/roundcubemail-plugins-kolab/tree/plugins/tasklist
Unfortunately they are using the Kolab custom storage engine instead of caldav, but the functionality is the same. I'm posting this for inspiration purposes.

Also, I wanted to see how the 2do app (iOS) implements subtasks since I'm using it on my phone with owncloud and it looks like the previously mentioned "related-to" property is used by this app too.
Here is the export of a sample calendar from 2do (with X-2DOAPP-METADATA content removed to save space):

BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Apple Inc.//Mac OS X 10.10.2//EN
CALSCALE:GREGORIAN
BEGIN:VTODO
STATUS:NEEDS-ACTION
CREATED:20150226T123307Z
UID:4e49347c61474a23a9aedecf6af55fc9
X-2DOAPP-SYNC-IMAGE:1
X-2DOAPP-SYNC-AUDIO:1
SUMMARY:Project 1
LAST-MODIFIED:20150226T123651Z
DTSTAMP:20150226T123651Z
X-2DOAPP-METADATA;...
X-2DOAPP-AUDIO-DURATION:0.000000
SEQUENCE:0
END:VTODO
BEGIN:VTODO
STATUS:NEEDS-ACTION
CREATED:20150226T123254Z
UID:d11eaf8dd5be481b849f497027fc49fb
X-2DOAPP-SYNC-IMAGE:1
X-2DOAPP-SYNC-AUDIO:1
SUMMARY:Task 1
LAST-MODIFIED:20150226T123653Z
DTSTAMP:20150226T123653Z
X-2DOAPP-METADATA;...
X-2DOAPP-AUDIO-DURATION:0.000000
SEQUENCE:0
END:VTODO
BEGIN:VTODO
X-2DOAPP-METADATA;...
DTSTAMP:20150226T123654Z
X-2DOAPP-SYNC-IMAGE:1
STATUS:NEEDS-ACTION
SEQUENCE:0
X-2DOAPP-AUDIO-DURATION:0.000000
X-2DOAPP-SYNC-AUDIO:1
SUMMARY:Subtask 1.2
LAST-MODIFIED:20150226T123654Z
RELATED-TO:01a3f66015924b4793bff2cf7ca7fe86
CREATED:20150226T123638Z
END:VTODO
BEGIN:VTODO
X-2DOAPP-METADATA;...
UID:59956c7e0798483e8c22cbd384796c74
DTSTAMP:20150226T123653Z
X-2DOAPP-SYNC-IMAGE:1
STATUS:NEEDS-ACTION
SEQUENCE:0
X-2DOAPP-AUDIO-DURATION:0.000000
X-2DOAPP-SYNC-AUDIO:1
SUMMARY:Project Task 1
LAST-MODIFIED:20150226T123653Z
RELATED-TO:4e49347c61474a23a9aedecf6af55fc9
CREATED:20150226T123401Z
END:VTODO
BEGIN:VTODO
STATUS:NEEDS-ACTION
CREATED:20150226T123501Z
UID:01a3f66015924b4793bff2cf7ca7fe86
X-2DOAPP-SYNC-IMAGE:1
X-2DOAPP-SYNC-AUDIO:1
SUMMARY:Task List 1
LAST-MODIFIED:20150226T123652Z
DTSTAMP:20150226T123652Z
X-2DOAPP-METADATA;...
X-2DOAPP-AUDIO-DURATION:0.000000
SEQUENCE:0
END:VTODO
BEGIN:VTODO
UID:63e6444ffad34fe4a61b6c3693625810
DTSTAMP:20150226T123654Z
X-2DOAPP-SYNC-IMAGE:1
STATUS:NEEDS-ACTION
SEQUENCE:0
X-2DOAPP-AUDIO-DURATION:0.000000
X-2DOAPP-SYNC-AUDIO:1
SUMMARY:Subtask 1.1
LAST-MODIFIED:20150226T123654Z
RELATED-TO:01a3f66015924b4793bff2cf7ca7fe86
CREATED:20150226T123643Z
END:VTODO
END:VCALENDAR

**Comment by xygrec** [comment]
PS: This is how Kolab Tasks looks like. (from Kolab.org)

![kolab-tasks-listing](https://cloud.githubusercontent.com/assets/11230927/6414462/9933be10-be65-11e4-83e8-cb908a4efadb.png)

**Comment by cmeeren** [comment]
I would love to see subtasks implemented. It is pretty much a must for me.

**Comment by sahne123** [comment]
For a simple Checklist using the description should do the trick. Saw this on the Android Tasks App. They are also using the CALDAV standard and stores it in a readable way for other clients:

```
BEGIN:VTODO
CREATED:20150618T083705Z
LAST-MODIFIED:20150618T083705Z
DESCRIPTION:Summary lorem ipsum my checklist\n[ ] Do something\n[ ] dont forget\n[x] i've already \ndone that\n[ ] important!
SUMMARY:test
PERCENT-COMPLETE:0
STATUS:NEEDS-ACTION
DTSTAMP:20150618T083819Z
UID:52eb683c-1c1a-4083-82d3-b53d84e1077f
END:VTODO
```
On other Clients the checklist will still be human readable:
```
Summary lorem ipsum my checklist
[ ] Do something
[ ] dont forget
[x] i've already done that
[ ] important!
```

**Comment by animalillo** [comment]
Kind of markdown support would be neat, and compatible with any client, altought not all would display the markdown with style

**Comment by raimund-schluessler** [comment]
An early implementation can be found here #202.

**Comment by DemoFreak** [comment]
Cool! :-)
I tried this immediately, but it seems that I'm doing sth wrong here. When I checkout the Subtasks branch, all my existing tasks are gone (same as when I checkout the todays master, BTW).
When I switch back to tag v0.6 all seems fine again.

- git checkout Subtasks
- reload apache2
-> all tasks gone

- git checkout v0.6
- reload apache2
-> all tasks re-appear

Do I have to do sth else but simply reload apache2?

**Comment by Henni** [comment]
@DemoFreak you are propably using owncloud 8.0
The current master and the subtasks branch require owncloud 8.1 which is not released yet.

**Comment by DemoFreak** [comment]
@Henni Indeed. Thx for hinting.

**Comment by raimund-schluessler** [comment]
Sorry, I should've said that.

**Comment by raimund-schluessler** [comment]
The subtasks branch has reached a state, where it could use some testing. Please report all issues found in #202. Thanks!

**Comment by AshleyYakeley** [comment]
There are different levels of subtask support...
* Level 0: All tasks are top level, there are no subtasks.
* Level 1: Top-level tasks may have subtasks. Subtasks may not have subtasks.
* Level 2: "Tree": every task has zero or one parent task, up to arbitrary depth. (Tasks without parent tasks are top-level tasks.)
* Level 3: "DAG": every task has zero or more parent tasks. (Cycles of parents are forbidden.) This allows diamond dependencies of tasks, for example, task A depends on tasks B & C, which both depend on task D. Task D is shown as "do-able", once it's done, then tasks B & C become "do-able".

I'm currently using "Getting Things Gnome", which is at level 3, the only task app I know at this level. It's great for people like me who like to break things down into tiny pieces.

**Comment by raimund-schluessler** [comment]
Following your classification, this app here is at level 2. Tasks only have one parent, but can have multiple subtasks which themselves may have subtasks.

As far as I understand the iCalendar specification, which the used CalDAV implementation relies on, having multiple parents is not supported (see http://tools.ietf.org/html/rfc5545#section-3.8.4.5). This means implementing level 3 is not possible without breaking the specification (which we don't want ;).

**Comment by AshleyYakeley** [comment]
Can't you just have multiple RELATED-TO properties, per section 3.1.2?

>Some properties defined in the iCalendar object can have multiple values.  The general rule for encoding multi-valued items is to simply create a new content line for each value, including the property name.

**Comment by raimund-schluessler** [comment]
I created a separate issue for your request in #235. Let's discuss your request there, as it's not a blocker but rather an enhancement for the current implementation of subtasks and I plan to close this issue here after #202 is merged.

**Comment by raimund-schluessler** [comment]
Implemented with #202.
